eclair-core/src/main/scala/fr/acinq/eclair/crypto/Sphinx.scala

```diff
@@ -74,298 +48,323 @@ object Sphinx extends Logging {
+        65
+      case _ =>
+        // The 1.1 BOLT spec changed the frame format to use variable-length per-hop payloads.
+        // The first bytes contain a varint encoding the length of the payload data (not including the trailing mac).
+        // Since messages are always smaller than 65535 bytes, this varint will either be 1 or 3 bytes long.
+        val lengthPrefix = payload.take(3)
+        val decodedLength = CommonCodecs.varintoverflow.decode(lengthPrefix.bits).require
+        val dataLength = decodedLength.value
+        val prefixLength = lengthPrefix.length - decodedLength.remainder.toByteVector.length
+        (prefixLength + dataLength + MacLength).toInt
     }
   }
 
   /**
-    * Compute the next packet from the current packet and node parameters.
-    * Packets are constructed in reverse order:
-    * - you first build the last packet
-    * - then you call makeNextPacket(...) until you've build the final onion packet that will be sent to the first node
-    * in the route
+    * Decrypting an onion packet yields a payload for the current node and the encrypted packet for the next node.
     *
-    * @param payload             payload for this packed
-    * @param associatedData      associated data
-    * @param ephemeralPublicKey ephemeral key for this packed
-    * @param sharedSecret        shared secret
-    * @param packet              current packet (1 + all zeroes if this is the last packet)
-    * @param routingInfoFiller   optional routing info filler, needed only when you're constructing the last packet
-    * @return the next packet
+    * @param payload      decrypted payload for this node.
+    * @param nextPacket   packet for the next node.
+    * @param sharedSecret shared secret for the sending node, which we will need to return failure messages.
     */
-  private def makeNextPacket(payload: ByteVector, associatedData: ByteVector32, ephemeralPublicKey: ByteVector, sharedSecret: ByteVector32, packet: Packet, routingInfoFiller: ByteVector = ByteVector.empty): Packet = {
-    require(payload.length == PayloadLength)
-
-    val nextRoutingInfo = {
-      val routingInfo1 = payload ++ packet.hmac ++ packet.routingInfo.dropRight(PayloadLength + MacLength)
-      val routingInfo2 = routingInfo1 xor generateStream(generateKey("rho", sharedSecret), MaxHops * (PayloadLength + MacLength))
-      routingInfo2.dropRight(routingInfoFiller.length) ++ routingInfoFiller
+  case class DecryptedPacket(payload: ByteVector, nextPacket: wire.OnionPacket, sharedSecret: ByteVector32) {
+
+    val isLastPacket: Boolean = payload.head match {
+      // In Bolt 1.0 the last hop is signaled via an empty hmac.
+      case 0 => nextPacket.hmac == ByteVector32.Zeroes
+      // In Bolt 1.1 the last hop can also be signaled via a dedicated TLV type with type=0x00.
+      case 0xfd => payload(3) == 0 || nextPacket.hmac == ByteVector32.Zeroes
+      case _ => payload(1) == 0 || nextPacket.hmac == ByteVector32.Zeroes
     }
```

Comment:
Is it possible to have `nextPacket.hmac == ByteVector32.Zeroes` and not be the last packet ?
Is it an error is `nextPacket.hmac == ByteVector32.Zeroes` but `payload(1) != 0` ?

Reply:
There is one chance in `2^256` that `ByteVector32.Zeroes` ends up being the correct mac value for a non-last packet, but that's already the case with the current scheme :)
Apart from that it shouldn't change what we have now, we just added another termination condition. It's unclear yet how it will be used, but we wanted to have that functionality ready to reclaim some space inside the onion (instead of wasting 32 bytes).